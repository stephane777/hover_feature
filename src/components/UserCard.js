import React from "react";
import { withHover } from "./withHover";
import PropTypes from "prop-types";
import { Button } from "react-bootstrap";
import UserForm from "./UserForm";

const UserCard = (props) => {
	const [loadMore, setLoadMore] = React.useState(false);
	const { hover, data } = props;
	const {
		name: { title, first, last },
		location: {
			city,
			country,
			street,
			postcode,
			timezone: { description },
		},
		dob: { age },
		registered: { date },
		email,
		phone,
		cell,
		picture: { thumbnail },
		login: { uuid, username },
	} = data;

	const tooltipStyle = {
		position: "absolute",
		top: "6rem",
		left: "0",
	};
	const userCardWrapper = {
		width: "100%",
		height: "100%",
		display: "flex",
		flexDirection: "column",
		justifyContent: "space-around",
		alignItems: "flex-start",
		overflow: "hidden",
		fontSize: "1.4rem",
	};
	const handleLoadMore = () => {
		setLoadMore(true);
	};
	return (
		<div id={uuid} style={userCardWrapper}>
			<div className="userCard__header">
				<div className="userCard__header_img">
					<img src={thumbnail} alt={`${title} ${first} ${last}`} />
				</div>
				<div className="userCard__header_details">
					<div>
						<span className="userCard__title">{`${title} ${first} ${last} ${age}`}</span>
					</div>
					<div className="bold">
						<span>{username}</span>
					</div>
					<div>
						<span>{email}</span>
					</div>
					<div>
						<span>{phone}</span>
					</div>
					<div>
						<span>{cell}</span>
					</div>
					{loadMore && (
						<React.Fragment>
							<div>
								<span>{`${street.number} ${street.name}`}</span>
							</div>
							<div>
								<span>{`${postcode} ${city} ${country}`}</span>
							</div>
							<div>
								<span>Timezone: {description}</span>
							</div>
							<div>
								<span>Registered: {date}</span>
							</div>
						</React.Fragment>
					)}
				</div>
			</div>
			{!loadMore && (
				<div className="userCard__more">
					<Button
						style={{ fontWeight: "bold" }}
						variant="primary"
						onClick={handleLoadMore}
					>
						Load more
					</Button>
				</div>
			)}
			{hover && !loadMore && (
				<div
					data-testid="userTooltip"
					className="userTooltip"
					style={tooltipStyle}
				>
					<div className="userTooltip__arrow"></div>
					<span className="userTooltip__content">{`${street.number} ${street.name}`}</span>
					<span className="userTooltip__content">{`${postcode} ${city} ${country}`}</span>
					<span className="userTooltip__content">Registered date :{date}</span>
				</div>
			)}
		</div>
	);
};

UserCard.propTypes = {
	hover: PropTypes.bool.isRequired,
	data: PropTypes.shape({
		cell: PropTypes.string,
		dob: PropTypes.object,
		email: PropTypes.string,
		gender: PropTypes.string,
		id: PropTypes.object,
		location: PropTypes.object,
		login: PropTypes.object,
		name: PropTypes.object,
		nat: PropTypes.string,
		phone: PropTypes.string,
		picture: PropTypes.object,
	}),
};

export default withHover(UserCard);
