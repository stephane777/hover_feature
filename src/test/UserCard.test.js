import React from "react";
import { render, screen, within, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import TestRenderer, { act } from "react-test-renderer";
import UserCard from "../components/UserCard";

const hover = false;
const data = {
	gender: "female",
	name: {
		title: "Mrs",
		first: "Emily",
		last: "Mortensen",
	},
	location: {
		street: {
			number: 8651,
			name: "Tingstedet",
		},
		city: "Randers Nv",
		state: "Nordjylland",
		country: "Denmark",
		postcode: 98755,
		coordinates: {
			latitude: "-29.6734",
			longitude: "110.0968",
		},
		timezone: {
			offset: "+9:00",
			description: "Tokyo, Seoul, Osaka, Sapporo, Yakutsk",
		},
	},
	email: "emily.mortensen@example.com",
	login: {
		uuid: "488edda9-f042-434f-8162-e4dddf7055f5",
		username: "blackwolf380",
		password: "gotyoass",
		salt: "aplnp8Vf",
		md5: "cef86433e959d301a61dae314a6261fc",
		sha1: "a1efa322b6074e5f8c28ae5e0cb59820cac629cc",
		sha256: "d84520956439ab48d1b9b90dfdcab9a833f50e254de5807f878a70a5e3a75653",
	},
	dob: {
		date: "1990-02-09T04:06:25.218Z",
		age: 31,
	},
	registered: {
		date: "2017-05-09T01:55:53.657Z",
		age: 4,
	},
	phone: "94033936",
	cell: "92381351",
	id: {
		name: "CPR",
		value: "090290-9172",
	},
	picture: {
		large: "https://randomuser.me/api/portraits/women/78.jpg",
		medium: "https://randomuser.me/api/portraits/med/women/78.jpg",
		thumbnail: "https://randomuser.me/api/portraits/thumb/women/78.jpg",
	},
	nat: "DK",
};

let props = { hover, data };

describe("test UserCard component with hover false", () => {
	it("should render the component", () => {
		const { container } = render(<UserCard {...props} />);
		expect(container.firstChild.className).toBe("userCard");
	});
	it("should show an image :", () => {
		render(<UserCard {...props} />);
		expect(screen.getByRole("img")).toHaveAttribute(
			"src",
			"https://randomuser.me/api/portraits/thumb/women/78.jpg"
		);
		expect(screen.getByAltText("Mrs Emily Mortensen"));
		expect(screen.getByRole("img")).toBeInTheDocument();
	});
	it("should show Gender, firstname, lastname, login, email, Phone and Cell", () => {
		render(<UserCard {...props} />);
		expect(screen.getByText("Mrs Emily Mortensen 31")).toBeInTheDocument();
		expect(screen.getByText("blackwolf380"));
		expect(screen.getByText("emily.mortensen@example.com")).toBeInTheDocument();
		expect(screen.getByText("94033936")).toBeInTheDocument();
		expect(screen.getByText("92381351")).toBeInTheDocument();
		expect(screen.getByText(/Load more/)).toBeInTheDocument();
	});
	it("shouldn't show the tooltip if not hovered", () => {
		props = {
			...props,
			hover: false,
		};
		expect(screen.queryByTestId("tooltip")).toBeNull();
	});
});
describe("test UserCard component with hover true", () => {
	it("should display the tooltip if hovered", () => {
		props = {
			...props,
			hover: true,
		};

		render(<UserCard {...props} />);
		expect(screen.getByText(/Load more/)).toBeInTheDocument();
		expect(screen.getByTestId("userTooltip")).toBeInTheDocument();

		const tooltipNode = document.querySelector(".userTooltip");
		const utils = within(tooltipNode);

		expect(utils.queryByText("Mrs Emily Mortensen 31")).toBeNull();
		expect(utils.getByText("8651 Tingstedet")).toBeInTheDocument();
		expect(utils.getByText("98755 Randers Nv Denmark")).toBeInTheDocument();
		expect(
			utils.getByText("Registered date :2017-05-09T01:55:53.657Z")
		).toBeInTheDocument();
	});
});

describe("Load more should display more data, remove the button and no tooltip", () => {
	it("should load more ", () => {
		props = {
			...props,
			hover: false,
		};
		const { rerender } = render(<UserCard {...props} />);
		expect(screen.getByText(/Load more/)).toBeInTheDocument();
		fireEvent.click(screen.getByText(/Load more/));
		expect(screen.queryByText(/Load more/)).toBeNull();

		expect(screen.getByText(/8651 Tingstedet/)).toBeInTheDocument();
		expect(screen.getByText(/98755 Randers Nv Denmark/)).toBeInTheDocument();
		expect(
			screen.getByText(/Timezone: Tokyo, Seoul, Osaka, Sapporo, Yakutsk/)
		).toBeInTheDocument();
		expect(
			screen.getByText(/Registered: 2017-05-09T01:55:53.657Z/)
		).toBeInTheDocument();
		props = {
			...props,
			hover: true,
		};
		rerender(<UserCard {...props} />);
		expect(screen.queryByTestId(/userTooltip/)).toBeNull();
	});
});

describe("test SNAPSHOT", () => {
	let root;
	it("should match the snapshot for hover false", () => {
		props = {
			...props,
			hover: false,
		};
		act(() => {
			root = TestRenderer.create(<UserCard {...props} />);
		});
		expect(root.toJSON()).toMatchSnapshot();
	});
	it("should match snapshot for hover true", () => {
		props = {
			...props,
			hover: true,
		};
		act(() => {
			root.update(<UserCard {...props} />);
		});
		expect(root.toJSON()).toMatchSnapshot();
	});
});
