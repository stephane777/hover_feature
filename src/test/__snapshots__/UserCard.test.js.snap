// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test SNAPSHOT should match snapshot for hover true 1`] = `
<div
  className="userCard"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    id="488edda9-f042-434f-8162-e4dddf7055f5"
    style={
      Object {
        "alignItems": "flex-start",
        "display": "flex",
        "flexDirection": "column",
        "fontSize": "1.4rem",
        "height": "100%",
        "justifyContent": "space-around",
        "overflow": "hidden",
        "width": "100%",
      }
    }
  >
    <div
      className="userCard__header"
    >
      <div
        className="userCard__header_img"
      >
        <img
          alt="Mrs Emily Mortensen"
          src="https://randomuser.me/api/portraits/thumb/women/78.jpg"
        />
      </div>
      <div
        className="userCard__header_details"
      >
        <div>
          <span
            className="userCard__title"
          >
            Mrs Emily Mortensen 31
          </span>
        </div>
        <div
          className="bold"
        >
          <span>
            blackwolf380
          </span>
        </div>
        <div>
          <span>
            emily.mortensen@example.com
          </span>
        </div>
        <div>
          <span>
            94033936
          </span>
        </div>
        <div>
          <span>
            92381351
          </span>
        </div>
      </div>
    </div>
    <div
      className="userCard__more"
    >
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "fontWeight": "bold",
          }
        }
        type="button"
      >
        Load more
      </button>
    </div>
    <div
      className="userTooltip"
      data-testid="userTooltip"
      style={
        Object {
          "left": "0",
          "position": "absolute",
          "top": "6rem",
        }
      }
    >
      <div
        className="userTooltip__arrow"
      />
      <span
        className="userTooltip__content"
      >
        8651 Tingstedet
      </span>
      <span
        className="userTooltip__content"
      >
        98755 Randers Nv Denmark
      </span>
      <span
        className="userTooltip__content"
      >
        Registered date :
        2017-05-09T01:55:53.657Z
      </span>
    </div>
  </div>
</div>
`;

exports[`test SNAPSHOT should match the snapshot for hover false 1`] = `
<div
  className="userCard"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    id="488edda9-f042-434f-8162-e4dddf7055f5"
    style={
      Object {
        "alignItems": "flex-start",
        "display": "flex",
        "flexDirection": "column",
        "fontSize": "1.4rem",
        "height": "100%",
        "justifyContent": "space-around",
        "overflow": "hidden",
        "width": "100%",
      }
    }
  >
    <div
      className="userCard__header"
    >
      <div
        className="userCard__header_img"
      >
        <img
          alt="Mrs Emily Mortensen"
          src="https://randomuser.me/api/portraits/thumb/women/78.jpg"
        />
      </div>
      <div
        className="userCard__header_details"
      >
        <div>
          <span
            className="userCard__title"
          >
            Mrs Emily Mortensen 31
          </span>
        </div>
        <div
          className="bold"
        >
          <span>
            blackwolf380
          </span>
        </div>
        <div>
          <span>
            emily.mortensen@example.com
          </span>
        </div>
        <div>
          <span>
            94033936
          </span>
        </div>
        <div>
          <span>
            92381351
          </span>
        </div>
      </div>
    </div>
    <div
      className="userCard__more"
    >
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "fontWeight": "bold",
          }
        }
        type="button"
      >
        Load more
      </button>
    </div>
  </div>
</div>
`;
